ports/vcpkg-downloads/download.json

```
@@ -0,0 +1,55 @@
+{
+  "downloads" : [
+    {
+      "output_variable" : "extracted_src_github",
+      "from": "github",
+      "repository" : "facebook/zstd",
+      "ref": "v1.5.6",
+      "sha512" : "ca12dffd86618ca008e1ecc79056c1129cb4e61668bf13a3cd5b2fa5c93bc9c92c80f64c1870c68b9c20009d9b3a834eac70db72242d5106125a1c53cccf8de8",
+      "head_ref": "dev",
+      "patches" : [],
+      "target_dir" : "${other_src_dir}/zstd"
+    },

```

Could also be:
```suggestion
    "extracted_src_github" : {
      "from": "github",
      "repository" : "facebook/zstd",
      "ref": "v1.5.6",
      "sha512" : "ca12dffd86618ca008e1ecc79056c1129cb4e61668bf13a3cd5b2fa5c93bc9c92c80f64c1870c68b9c20009d9b3a834eac70db72242d5106125a1c53cccf8de8",
      "head_ref": "dev",
      "patches" : [],
      "target_dir" : "${other_src_dir}/zstd"
    },
```

Basically dropping the `output_variable` field

---

I'm in favor of this because otherwise the set of allowed fields would depend on the value of `from` which is way harder to parse compared to `"from": {...}`.

---

You suggested here to make the output variable the key of the object instead of having an array of keyless objects. I propose to make the value of `from` the key:

```json
"github" : {
      "output-variable": "extracted_src_github",
      "repository" : "facebook/zstd",
      "ref": "v1.5.6",
      "sha512" : "ca12dffd86618ca008e1ecc79056c1129cb4e61668bf13a3cd5b2fa5c93bc9c92c80f64c1870c68b9c20009d9b3a834eac70db72242d5106125a1c53cccf8de8",
      "head_ref": "dev",
      "patches" : [],
      "target_dir" : "${other_src_dir}/zstd"
    }
```

This way you can decide more easily the download method.

---

I think the main benefit for that is schema validation, however you might end up with something like:

```
[
 "github" : { .... },
 "github" : { .... },
 "github" : { .... },
 "github" : { .... },
 "gitlab" : { .... },
 "git" : { .... },
]
```

Leading to `something.0.github.--- something.1.github.--- something.2.github.---`. The is no meaning in that selection if there is duplicates. 

While:
```
[
 "src1" : { .... },
 "src2" : { .... },
 "src3" : { .... },
 "src4" : { .... },
]
```
Actually has meaning since `srcX` has to be an unique identifier since otherwise outputs get overwritten. I also see the json keys as a right to left assignment in this case (data on the right creates the key on the left).

---

> Leading to `something.0.github.--- something.1.github.--- something.2.github.---`. The is no meaning in that selection if there is duplicates.

Why is this a problem? To my understanding, JSON arrays are allowed to have multiple elements (objects) with the same key, e.g. `"{"key":  "value", "key":  "value"}` is not allowed whereas `["key":  {"subkey": "value"}, "key":  {"subkey": "value"}]` is allowed.

> data on the right creates the key on the left

The output variable itself is data. According to this, it should be on the right whereas `github` or `gitlab` are less data-ish. I think JSON should go from general to specific. The output variable is a lot more specific than `github` or `gitlab`.

---

> `["key":  {"subkey": "value"}, "key":  {"subkey": "value"}]`

This is no valid JSON...

---

would need to be `[ { "key":  {"subkey": "value"} }, { "key":  {"subkey": "value"} } ]`

---

I prefer the `output_variable` as key variant, because you prevent collisions by design with that.